mlflow/server/auth/__init__.py

```diff
@@ -8,15 +8,51 @@
 """
 
 import logging
+from flask import Flask, request, make_response, Response, redirect, flash, render_template_string
 
 from mlflow.server import app
-
+from mlflow.server.auth import store
+from mlflow.server.auth.config import app_config
 
 _logger = logging.getLogger(__name__)
 
 
+class ROUTES:
+    USERS = "/mlflow/users"
+    SIGNUP = "/mlflow/signup"
+
+
+UNPROTECTED_ROUTES = [ROUTES.USERS, ROUTES.SIGNUP]
+
+
+def is_unprotected_route(path: str) -> bool:
+    if path.startswith("/static"):
+        return True
+    return path in UNPROTECTED_ROUTES
+
+
+def make_basic_auth_response() -> Response:
+    res = make_response()
+    res.status_code = 401
+    res.headers["WWW-Authenticate"] = 'Basic realm="mlflow"'
+    return res
+
+
 def _before_request():
-    # TODO: Implement authentication
+    _user = request.authorization.username if request.authorization else None
+    _logger.info(f"before_request: {request.method} {request.path} (user: {_user})")
+    if is_unprotected_route(request.path):
+        return
+
+    if request.authorization is None:
+        return make_basic_auth_response()
+
+    username = request.authorization.username
+    password = request.authorization.password
+    if not store.authenticate_user(username, password):
+        # let user attempt login again
+        return make_basic_auth_response()
```

Comment:
let user attempt login if the username / password is wrong

not sure how we can flash the user with message on this